Find Issues command

Burhan-Q opened the pull request:
# Rationale

<!-- Explain why you are making this change. Describe the problem. -->

Brief exchange with SteveO on the [docs repo](https://github.com/voxel51/voxel51-docs/issues/23#issuecomment-2659519921) made me think this could be a good addition.

## Changes

<!-- Describe the changes. -->

Adds Discord `/ref_issue` command, that allows users to search for 5 relevant issues. Includes two arguments `repo` and `query` to allow specifying the repo to search and the keywords in title/topic of Issues to find.

## Testing

<!-- Describe the way the changes were tested. -->

😆 unfortunately not at all. Did my best to mirror what other sections of code were doing and to make sure it had a logical flow.  Not sure if the "default" behavior included (try to search `voxel51/fiftyone` if no repo is specified).

<!-- Optional Sections:

## Screenshots
## To Do
## Notes
## Related

-->

<!-- Template for collapsed sections
<details>
<summary></summary>
</details>
-->

Burhan-Q committed:
add issue query

Burhan-Q committed:
Create find_issues.py

Burhan-Q committed:
revert easy_github_contribution.py

Burhan-Q committed:
default when `repo` is `None`

Burhan-Q committed:
add missing `repo`

afoley587 commented on voxelbot/plugins/find_issues.py:
```diff
@@ -0,0 +1,42 @@
+        await client.message_create(event.channel, "An error occurred while searching for issues. Please try again later.")
+
+
+@repo_metrics.autocomplete("repo")
```
Where is this defined / imported from?

Burhan-Q replied:
This was a bit of copy-code that didn't get updated. This has been changed to `@ref_issue.autocomplete("repo")` which mirrors the autocomplete definition from the `repo_metrics` function.

afoley587 commented on voxelbot/plugins/find_issues.py:
```diff
@@ -0,0 +1,42 @@
+    issue_links = "\n".join(f"{issue.title}: {issue.html_url}" for issue in issues[:5])
+    await client.message_create(event.channel, f"Top relevant issues:\n{issue_links}")
+    
+    except GithubException as e:
```
This isn't imported so will throw an error

afoley587 commented on voxelbot/plugins/find_issues.py:
```diff
@@ -0,0 +1,42 @@
+@ALL.interactions(is_global=True, wait_for_acknowledgement=True)
+async def ref_issue(client, event, query: ("str", "Topic to search"), repo: ("str", "Repo to search.")):  # type: ignore # noqa: F722
+    """Provides specific metric(s) for the voxel51/fiftyone repository."""
+    if repo is not None and repo not in (VOXEL51_REPOS):
```
Why are we making `VOXEL51_REPOS` a tuple here, I believe it's already a list?

Burhan-Q replied:
The `( )` were leftovers and have been removed.

Burhan-Q committed:
PR comment changes

afoley587 commented on voxelbot/plugins/find_issues.py:
```diff
@@ -0,0 +1,42 @@
+    issue_links = "\n".join(f"{issue.title}: {issue.html_url}" for issue in issues[:5])
+    await client.message_create(event.channel, f"Top relevant issues:\n{issue_links}")
+    
+    except GithubException as e:
```
You have this `except` here, but no `try`, so this isn't valid python. You would need a `try` somewhere else.

afoley587 commented on voxelbot/plugins/find_issues.py:
```diff
@@ -0,0 +1,42 @@
+    issues = repo.search_issues(f'{query} in:title,body')
+    
+    if issues.totalCount == 0:
+            return await client.message_create(message.channel, "No relevant issues found.")
```
`message` doesn't seem to be defined anywhere. Spacing also seems off.

Burhan-Q replied:
You're right. I've updated this to use `event.channel` which I'm thinking is the correct way to go based on the [GitHub contributions plugin](https://github.com/voxel51/VoxelBot/blob/8b4e284b12a0a2932878103c8fd79ceda631cef4/voxelbot/plugins/easy_github_contribution.py#L102) since I _think_ that the `event` object should be the same.

afoley587 reviewed (commented):
How is something like?

```python
.
.
from hata import Client, ClientWrapper, Message
.
.
.
@ALL.interactions(is_global=True, wait_for_acknowledgement=True)
async def ref_issue(client: Client, message: Message, query: ("str", "Topic to search"), repo: ("str", "Repo to search.")):  # type: ignore # noqa: F722
    """Provides specific metric(s) for the voxel51/fiftyone repository."""
    if repo is not None and repo not in VOXEL51_REPOS:
        abort("Must be Voxel51 repo.")
    elif repo is None:
        repo = "voxel51/fiftyone"  # default
    
    try:
        GITHUB.get_repo(repo)
        issues = repo.search_issues(f'{query} in:title,body')
        
        if issues.totalCount == 0:
            return await client.message_create(message.channel, "No relevant issues found.")
        
        issue_links = "\n".join(f"{issue.title}: {issue.html_url}" for issue in issues[:5])
        await client.message_create(message.channel, f"Top relevant issues:\n{issue_links}")
    
    except GithubException as e:
        logging.error(f"GitHub API error: {e}")
        await client.message_create(message.channel, "An error occurred while searching for issues. Please try again later.")
```

afoley587 commented on voxelbot/plugins/find_issues.py:
```diff
@@ -0,0 +1,42 @@
+ALL = ClientWrapper()
+
+GITHUB = Github(auth=Auth.Token(GITHUB_TOKEN))
+VOXEL51_REPOS = [r.name for r in GITHUB.get_repos(type="public")]
```
It looks like `type` isn't a keyword. But, `visibility` is:

```
>>> VOXEL51_REPOS = [r.name for r in GITHUB.get_repos(type="public")]
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: Github.get_repos() got an unexpected keyword argument 'type'
>>> GITHUB.get_repos
<bound method Github.get_repos of <github.MainClass.Github object at 0x105137f70>>
>>> help(GITHUB.get_repos
... 
KeyboardInterrupt
>>> help(GITHUB.get_repos)

>>> VOXEL51_REPOS = [r.name for r in GITHUB.get_repos(visibility="public")]
```

afoley587 replied:
This also gets ALL public repos in ALL of github. Is there a function to only get ours?

Burhan-Q committed:
add GitHub org

Burhan-Q committed:
add try-except block, fix spacing

Burhan-Q committed:
use `event.channel` for responding to user command

Burhan-Q committed:
use `r.full_name` instead of `r.name` to get `owner/repo`

Burhan-Q commented:
Since I _think_ that this will work as a slash command `/` (based on what I saw in the other plugins), I stuck with `event` and just used `event.channel` which I believe should be a [InteractionEvent](https://www.astil.dev/project/hata/docs/hata/discord/interaction/interaction_event/interaction_event/InteractionEvent) object, but to be honest, I haven't used `hata` before, so I'm trying to figure this out on the fly.

I added a line for fetching the GitHub organization to limit the ones that are listed. I choose to return the `full_name` attribute so that it would return `owner/repo` which should be directly compatible with the `get_repo` method.

afoley587 commented on voxelbot/plugins/find_issues.py:
```diff
@@ -0,0 +1,44 @@
+        repo = "voxel51/fiftyone"  # default
+    
+    try:
+        GITHUB.get_repo(repo)
```
do you need to do `repo = GITHUB.get_repo(repo)`?

Burhan-Q committed:
fix docstring

afoley587 commented on voxelbot/plugins/find_issues.py:
```diff
@@ -0,0 +1,44 @@
+    
+    try:
+        GITHUB.get_repo(repo)
+        issues = repo.search_issues(f'{query} in:title,body')
```
```
>>> repo = GITHUB.get_repo('voxel51/fiftyone')
>>> issues = repo.search_issues()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: 'Repository' object has no attribute 'search_issues'
```

But, 
```
issues = GITHUB.search_issues(f"{query} repo:{repo.full_name} in:title,body")
```
seems to work.

afoley587 replied:
Also, 

```python
github.GithubException.GithubException: Query must include 'is:issue' or 'is:pull-request': 422 {"message": "Query must include 'is:issue' or 'is:pull-request'", "documentation_url": "https://docs.github.com/rest/search/search#search-issues-and-pull-requests", "status": "422"}
```

So maybe change `search_issues(f"{query} repo:{repo.full_name} in:title,body")` to `search_issues(f"{query} repo:{repo.full_name} in:title,body is:issue")`

Burhan-Q committed:
changes per PR comments

Burhan-Q commented on voxelbot/plugins/find_issues.py:
```diff
@@ -0,0 +1,44 @@
+        repo = "voxel51/fiftyone"  # default
+    
+    try:
+        repo = GITHUB.get_repo(repo)
+        issues = GITHUB.search_issues(f"{query} repo:{repo.full_name} in:title,body")
```
```diff
- repo = GITHUB.get_repo(repo)
- issues = GITHUB.search_issues(f"{query} repo:{repo.full_name} in:title,body")
+ issues = GITHUB.search_issues(f"{query} repo:{repo} in:title,body")
```

afoley587 replied:
```
github.GithubException.GithubException: Query must include 'is:issue' or 'is:pull-request': 422 {"message": "Query must include 'is:issue' or 'is:pull-request'", "documentation_url": "https://docs.github.com/rest/search/search#search-issues-and-pull-requests", "status": "422"}
```

Should we add `is:issue`?

```python
        issues = GITHUB.search_issues(f"{query} repo:{repo.full_name} in:title,body")
```

to 

```python
        issues = GITHUB.search_issues(f"{query} repo:{repo.full_name} in:title,body is:issue")
```

Burhan-Q replied:
I wonder if this would make more sense now? Guessing that `voxel51/{repo}` would be the only format that would pass the conditional, so I would suspect it could be passed directly into the search.

Burhan-Q replied:
Added `is: issue` to the query. What do you think about removing the `get_repo()` call and passing the `repo` string directly into the query?

afoley587 replied:
That sounds like a fine solution to me!

Burhan-Q committed:
added 'is:issue' to query

afoley587 commented on voxelbot/plugins/find_issues.py:
```diff
@@ -0,0 +1,44 @@
+
+
+@ALL.interactions(is_global=True, wait_for_acknowledgement=True)
+async def ref_issue(client, event, query: ("str", "Topic to search"), repo: ("str", "Repo to search.")):  # type: ignore # noqa: F722
```
```suggestion
async def ref_issue(client: Client, event: Message, query: ("str", "Topic to search"), repo: ("str", "Repo to search.")):  # type: ignore # noqa: F722
```

Burhan-Q replied:
Forgive my ignorance, but I thought that `event` would be a `InteractionEvent` (like in [this example from the `hata` guide](https://www.astil.dev/project/hata/guides/slash#command-parameters)). So I would _think_ it should be:

```py
ref_issue(client: Client, event: InteractionEvent
```

since it's not waiting on a message, but a slash-command? I'm all for type-annotations, it wasn't clear to me what the practice here was, which is why I left it out originally. If `Message` is the correct type, I have no qualms with updating the argument and type-hint, but I was thinking that since I used the `faq` and `code-metrics` "triggers" that this would use `event` instead of `message`.

afoley587 replied:
You're probably right. I just think we should type hint - so you can change my suggestion to use the interaction events!

Burhan-Q replied:
100% agree with type hints! Updated :+1:

afoley587 commented on voxelbot/plugins/find_issues.py:
```diff
@@ -0,0 +1,44 @@
+import logging
+
+from github import Auth, Github, GithubException
+from hata import Client, ClientWrapper
```
```suggestion
from hata import Client, ClientWrapper, Message
```

Burhan-Q committed:
add type hints 🚀

Burhan-Q committed:
remove extraneous `get_repo` call

afoley587 commented on voxelbot/plugins/find_issues.py:
```diff
@@ -0,0 +1,43 @@
+
+
+@ALL.interactions(is_global=True, wait_for_acknowledgement=True)
+async ref_issue(client: Client, event: InteractionEvent, query: ("str", "Topic to search"), repo: ("str", "Repo to search.")):  # type: ignore # noqa: F722
```
This needs to be `async def`...

afoley587 replied:
Also, down below, you check if `repo is None`. But, with this method signature, repo will never be None. Please remove that check or, if you want a default, add it as a default arg

Burhan-Q committed:
include `def`, expand docstring, format line-length

Burhan-Q committed:
type hints, imports, docstrings, line formatting, and remove redundant conditional

Burhan-Q commented:
I appreciate the patience here, I've been doing a lot of this from mobile and _finally_ had the opportunity to pull this onto a local machine. Updated with type hints, docstrings, imports, and formatting (assumed line line of 100, `pylint` default, since there didn't seem to be a configured value). 

Added the `def` keyword for the function definition, and removed the redundant `None` checks. Also ran `pre-commit` locally.

I did notice that on my Windows machine, `inotify-simple` and `cchardet` failed to build. I reran without building and temporarily removing `inotify-simple` to install all packages, since I didn't find a "quick" solution. Not that it's critical for me at this juncture, but just wanted to point this out.

afoley587 commented:
Great, thanks for the contribution. I will merge which will trigger a deployment.

afoley587 commented:
@Burhan-Q issue with loading this plugin:

```
ValueError: `description` length is out of the expected range [2:100], got 104; "Returns a list of top-five related issues from a specified Voxel51 repository based on the user's query.".
```

Burhan-Q commented:
🤔 hm, guess the docstring needs shortening. Hadn't considered there would be a limit there.
